main/tests/test_variant.h

```diff
@@ -0,0 +1,77 @@
+/*************************************************************************/
+/* Copyright (c) 2007-2020 Juan Linietsky, Ariel Manzur.                 */
+/* Copyright (c) 2014-2020 Godot Engine contributors (cf. AUTHORS.md).   */
+/*                                                                       */
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef TEST_VARIANT_H
+#define TEST_VARIANT_H
+
+#include "core/variant.h"
+#include "core/variant_parser.h"
+
+#include "thirdparty/doctest/doctest.h"
+
+namespace TestVariant {
+
+TEST_CASE("[Variant] Writer and parser overflow check") {
+	// Writer.
+	int64_t a32 = 2147483648; // 2^31, so out of bounds for 32-bit signed int [-2^31,-2^31-1].
+	String a32_str;
+	VariantWriter::write_to_string(a32, a32_str);
+
+	CHECK(a32_str != "-2147483648"); // Should not wrap around.
+
+	int64_t b64 = 9223372036854775807; // 2^63-1, upper bound for signed 64-bit int.
+	String b64_str;
+	VariantWriter::write_to_string(b64, b64_str);
+
+	CHECK(b64_str == "9223372036854775807"); // Should not wrap around.
+
+	// Parser.
+	VariantParser::StreamString ss;
+	String errs;
+	int line;
+	Variant b64_parsed;
+	int64_t b64_int_parsed;
+
+	ss.s = b64_str;
+	VariantParser::parse(&ss, b64_parsed, errs, line);
+	b64_int_parsed = b64_parsed;
+
+	CHECK(b64_int_parsed == 9223372036854775807);
+
+	ss.s = "9223372036854775808"; // Overflowed by one.
+	VariantParser::parse(&ss, b64_parsed, errs, line);
+	b64_int_parsed = b64_parsed;
+
+	CHECK(b64_int_parsed == 9223372036854775807); // The result is clamped to max value.
```

Comment:
This one is interesting, I thought it would overflow but it just clamps, worth documenting I think.